src/Sulu/Bundle/PreviewBundle/Preview/Renderer/PreviewRenderer.php

```diff
@@ -0,0 +1,155 @@
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Bundle\PreviewBundle\Preview\Renderer;
+
+use Sulu\Bundle\PreviewBundle\Preview\Events;
+use Sulu\Bundle\PreviewBundle\Preview\Events\PreRenderEvent;
+use Sulu\Bundle\RouteBundle\Routing\Defaults\RouteDefaultsProviderInterface;
+use Sulu\Component\Webspace\Analyzer\Attributes\RequestAttributes;
+use Sulu\Component\Webspace\Manager\WebspaceManagerInterface;
+use Sulu\Component\Webspace\PortalInformation;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\HttpKernel\HttpKernelInterface;
+
+/**
+ * Renders preview responses.
+ */
+class PreviewRenderer implements PreviewRendererInterface
+{
+    /**
+     * @var RouteDefaultsProviderInterface
+     */
+    private $routeDefaultsProvider;
+
+    /**
+     * @var RequestStack
+     */
+    private $requestStack;
+
+    /**
+     * @var HttpKernelInterface
+     */
+    private $httpKernel;
+
+    /**
+     * @var WebspaceManagerInterface
+     */
+    private $webspaceManager;
+
+    /**
+     * @var EventDispatcherInterface
+     */
+    private $eventDispatcher;
+
+    /**
+     * @var array
+     */
+    private $previewDefaults;
+
+    /**
+     * @var string
+     */
+    private $environment;
+
+    /**
+     * @param RouteDefaultsProviderInterface $routeDefaultsProvider
+     * @param RequestStack $requestStack
+     * @param HttpKernelInterface $httpKernel
+     * @param WebspaceManagerInterface $webspaceManager
+     * @param EventDispatcherInterface $eventDispatcher
+     * @param array $previewDefaults
+     * @param string $environment
+     */
+    public function __construct(
+        RouteDefaultsProviderInterface $routeDefaultsProvider,
+        RequestStack $requestStack,
+        HttpKernelInterface $httpKernel,
+        WebspaceManagerInterface $webspaceManager,
+        EventDispatcherInterface $eventDispatcher,
+        array $previewDefaults,
+        $environment
+    ) {
+        $this->routeDefaultsProvider = $routeDefaultsProvider;
+        $this->requestStack = $requestStack;
+        $this->httpKernel = $httpKernel;
+        $this->webspaceManager = $webspaceManager;
+        $this->eventDispatcher = $eventDispatcher;
+        $this->previewDefaults = $previewDefaults;
+        $this->environment = $environment;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function render($object, $id, $webspaceKey, $locale, $partial = false)
+    {
+        $portalInformations = $this->webspaceManager->findPortalInformationsByWebspaceKeyAndLocale(
+            $webspaceKey,
+            $locale,
+            $this->environment
+        );
+
+        if (count($portalInformations) === 0 || !$this->routeDefaultsProvider->supports(get_class($object))) {
+            // TODO template for non renderable response
```

Comment:
@chirimoya , @danrot or any another i would prefer to add this in another PR when this is merged.